website/docs/Cache_Library_User_Guides/FDP_enabled_Cache.md

```diff
@@ -0,0 +1,92 @@
+
+### How does CacheLib use FDP?
+CacheLib's BigHash and BlockCache produce distinct IO patterns on the SSD. BigHash generates a random write pattern while BlockCache generates a sequential write pattern. In a conventional SSD, these writes get mixed up in the physical NAND media. This intermixing can lead to a higher SSD Write Amplification Factor (WAF). To combat this in production environments CacheLib uses upto [50% of the SSD as host over-provisioning](https://www.usenix.org/system/files/osdi20-berg.pdf). The Flexible Data Placement (FDP) support within CacheLib aims to segregate the BigHash and BlockCache data of CacheLib within the SSDs. This reduces the device WAF even when configured with 0% host over-provisioning and improves device endurance. FDP support within Navy is optional.
+
+| ![](alternate_fdp_navy.png) |
+|:--:|
+| *CacheLib IO Flow with and without FDP* |
+
+Since FDP directives are not yet supported by the Linux kernel block layer interface, we have used Linux kernel [I/O Passthru](https://www.usenix.org/system/files/fast24-joshi.pdf) mechanism (which leverages [io_uring_cmd interface](https://www.usenix.org/system/files/fast24-joshi.pdf)) as seen in the figure. 
+
+### FDP support within Navy
+The `FdpNvme` class embeds all the FDP related semantics and APIs and is used  by the `FileDevice` class. This can be extended to other modules in the future if they desire to use FDP support. The below are some key functions added to the `FdpNvme` class related to FDP and iouring_cmd.
+ 
+```cpp
+
+  // Allocates an FDP specific placement handle to modules using FdpNvme like Block Cache and Big Hash.
+  // This handle will be interpreted by the device for data placement.
+  int allocateFdpHandle();
+
+  // Prepares the Uring_Cmd sqe for read/write command with FDP directives.
+  void prepFdpUringCmdSqe(struct io_uring_sqe& sqe,
+                          void* buf,
+                          size_t size,
+                          off_t start,
+                          uint8_t opcode,
+                          uint8_t dtype,
+                          uint16_t dspec);
+```
+
+### When should FDP be enabled?
+When using Navy, FDP can help play a role in improving SSD endurance in comparison to Non-FDP. If the workload has both small objects and large objects, FDP's WAF gains will most likely be evident because FDP segregates these two in the SSD. In cases like CacheLib's CDN workload where BigHash is typically not configured, this FDP based segregation will make no difference to SSD WAF. On the other hand with CacheLib's KV Cache workload where both Big Hash and Block Cache are configured, we see the gains from using FDP.  
```

Comment:
Can you add some numbers here from your experiments in KV Cache workloads FDP vs without?

Reply:
Yes, we'll add the data we have from our KV Cache experiments with and without FDP.

Reply:
We've updated the results in the commit. Please check and let us know.

Reply:
appreciate it!

Reply:
With respect to configs:
We leave out configs in this description because we don't change the default configs obtained when downloading the traces. We expect the user to follow the steps mentioned in the "[Evaluating SSD hardware for Facebook workloads](https://cachelib.org/docs/Cache_Library_User_Guides/Cachebench_FB_HW_eval/)" page to [run the production traces with CacheBench](https://cachelib.org/docs/Cache_Library_User_Guides/Cachebench_FB_HW_eval/#running-cachebench-with-the-trace-workload). The only additional steps to enable FDP would be to use the flags mentioned in this page.

With respect to the specific trace used:
As mentioned the benefits of FDP can be seen with KV Cache traces and these benefits won't be seen with CDN or Storage traces because they don't configure the SOC.
There are multiple KV Cache traces in "s3://cachelib-workload-sharing/pub/kvcache/". Any of them can be used to observe the benefits. Again, this is why we don't point to a specific KV Cache trace in this description. 

However, we can update the PR to point to the particular KV Cache trace used in the results section presented for the sake of reproducibility.